roofit/roofitcore/src/TestStatistics/likelihood_builders.cxx

```diff
@@ -0,0 +1,325 @@
+      }
+      std::string name("likelihood for pdf ");
+      name += pdf->GetName();
+      subsidiary_likelihood = std::make_unique<RooSubsidiaryL>(name, allConstraints,
+                                                               (global_observables.set.getSize() > 0) ? global_observables.set : constrained_parameters.set);
+   }
+
+   // END CONSTRAINT COLLECTION; copied from RooAbsPdf::createNLL
+
+   return subsidiary_likelihood;
+}
+
+
+std::shared_ptr<RooAbsL>
+buildSimultaneousLikelihood(RooAbsPdf *pdf, RooAbsData *data, RooAbsL::Extended extended,
+                                                       ConstrainedParameters constrained_parameters, ExternalConstraints external_constraints,
+                                                       GlobalObservables global_observables, std::string global_observables_tag) {
+   auto sim_pdf = dynamic_cast<RooSimultaneous *>(pdf);
+   if (sim_pdf == nullptr) {
+      throw std::logic_error("Can only build RooSumL from RooSimultaneous pdf!");
+   }
+
+   // the rest of this function is an adaptation of RooAbsTestStatistic::initSimMode:
+
+   RooAbsCategoryLValue &simCat = (RooAbsCategoryLValue &)sim_pdf->indexCat();
+
+   // FIXME: process_empty_data_sets was previously member function processEmptyDataSets(), which returned extended_ for RooSumL (previously named RooSimultaneousL). See RooAbsL and RooSumL/RooUnbinnedL implementations. I think this should work here, but be careful in other subclasses!
+   bool process_empty_data_sets = RooAbsL::isExtendedHelper(pdf, extended);
+
+   TString simCatName(simCat.GetName());
+   // Note: important not to use cloned dataset here (possible when this code is run in Roo[...]L ctor), use the
+   // original one (which is data_ in Roo[...]L ctors, but data here)
+   std::unique_ptr<TList> dsetList{data->split(simCat, process_empty_data_sets)};
+   if (!dsetList) {
+      throw std::logic_error("buildSimultaneousLikelihood ERROR, index category of simultaneous pdf is missing in dataset, aborting");
+   }
+
+   // Count number of used states
+   std::size_t N_components = 0;
+
+   RooCatType *type;
+   std::unique_ptr<TIterator> catIter{simCat.typeIterator()};
+   while ((type = (RooCatType *)catIter->Next())) {
+      // Retrieve the PDF for this simCat state
+      RooAbsPdf *component_pdf = sim_pdf->getPdf(type->GetName());
+      auto dset = (RooAbsData *)dsetList->FindObject(type->GetName());
+
+      if (component_pdf && dset && (0. != dset->sumEntries() || process_empty_data_sets)) {
+         ++N_components;
+      }
+   }
+
+   // Allocate arrays
+   std::vector<std::unique_ptr<RooAbsL>> components;
+   components.reserve(N_components);
+   //   _gofSplitMode.resize(N_components);  // not used, Hybrid mode only, see below
+
+   // Create array of regular fit contexts, containing subset of data and single fitCat PDF
+   catIter->Reset();
```

Comment:
This caused enough pain. Please don't reuse iterators. Range-based for all the way.

Reply:
Sorry for the pain ;)